community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/ReversedSingleFileTransactionCursor.java

```diff
@@ -0,0 +1,177 @@
+ * per chunk "move". This is all internal, so from the outside it simply reverses a transaction log.
+ * The memory overhead compared to reading a log in the natural order is almost negligible.
+ *
+ * This cursor currently only works for a single log file, such that the given {@link ReadAheadLogChannel} should not be
+ * instantiated with a {@link LogVersionBridge} moving it over to other versions when exhausted. For reversing a whole
+ * log stream consisting of multiple log files have a look at {@link ReversedMultiFileTransactionCursor}.
+ *
+ * @see ReversedMultiFileTransactionCursor
+ */
+class ReversedSingleFileTransactionCursor implements TransactionCursor
+{
+    // Should this be passed in or extracted from the read-ahead channel instead?
+    private static final int CHUNK_SIZE = ReadAheadChannel.DEFAULT_READ_AHEAD_SIZE;
+
+    private final ReadAheadLogChannel channel;
+    private final TransactionCursor transactionCursor;
+    // Should be generally large enough to hold transactions in a chunk, where one chunk is the read-ahead size of ReadAheadLogChannel
+    private final ArrayCollection<CommittedTransactionRepresentation> chunkTransactions = new ArrayCollection<>( 20 );
+    // May be longer than required, offsetLength holds the actual length.
+    private final long[] offsets;
+    private int offsetsLength;
+    private int chunkStartOffsetIndex;
+    private int chunkLength;
+    private int chunkCursor;
+    private long totalSize;
+
+    ReversedSingleFileTransactionCursor( ReadAheadLogChannel channel, LogEntryReader<ReadableClosablePositionAwareChannel> logEntryReader )
+            throws IOException
+    {
+        this.channel = channel;
+        // There's an assumption here: that the underlying channel can move in between calls and that the
+        // transaction cursor will just happily read from the new position.
+        this.transactionCursor = new PhysicalTransactionCursor<>( channel, logEntryReader );
+        this.offsets = sketchOutTransactionStartOffsets();
+    }
+
+    // Also initializes offset indexes
+    // This method could use some way of reading log entries w/o creating objects. Would be great
+    private long[] sketchOutTransactionStartOffsets() throws IOException
+    {
+        // Grows on demand. Initially sized to be able to hold all transaction start offsets for a single log file
+        long[] offsets = new long[10_000];
+        int offsetCursor = 0;
+
+        long startOffset = channel.position();
+        while ( transactionCursor.next() )
+        {
+            if ( offsetCursor == offsets.length )
+            {   // Grow
+                offsets = Arrays.copyOf( offsets, offsetCursor * 2 );
+            }
+            offsets[offsetCursor++] = startOffset;
+            long nextStartOffset = channel.position();
+            assert nextStartOffset > startOffset : "It looks like this channel moved over to another version";
```

Comment:
That check is just an assertion to spot cases where you may have passed in a channel with some log version bridge so that it traverses multiple log versions. The check isn't fully covering tho, but I don't think it's invalid.

Reply:
I could imagine a couple of perverse cases with strange log rotation settings that would cause this to fail. Checking the log version on the channel seems like a wonderful idea :+1: